README.md
# aviauth-api
[![Release Version](https://img.shields.io/github/v/release/julleks/aviauth-api.svg?sort=semver)](https://github.com/julleks/aviauth-api/releases)
[![License: GPL v3](https://img.shields.io/badge/License-GPLv3-blue.svg)](https://github.com/julleks/aviauth-api/blob/develop/LICENSE)
[![CI](https://github.com/julleks/aviauth-api/actions/workflows/test.yml/badge.svg?branch=master)](https://github.com/julleks/aviauth-api/actions?query=branch%3A+master)
[![codecov](https://codecov.io/gh/julleks/aviauth-api/branch/master/graph/badge.svg)](https://codecov.io/gh/julleks/aviauth-api)

Authentication microservice based on FastAPI


# Installation guide

Clone the repository
```shell
git clone https://github.com/julleks/aviauth-api/
```

Install [pyenv](https://github.com/pyenv/pyenv/wiki#suggested-build-environment) and dependencies (macOS)
```shell
brew install openssl readline sqlite3 xz zlib

curl https://pyenv.run | bash

pyenv install 3.10.1
```

Install [poetry](https://python-poetry.org)
```shell
curl -sSL https://raw.githubusercontent.com/python-poetry/poetry/master/get-poetry.py | python -

source $HOME/.poetry/env
```

```shell
cd aviauth-api

pyenv local 3.10.1

poetry install

peotry shell
```

Install pre-commit if you are going to make commits to repository
```shell
pre-commit install
```

Set debug to `True` to disable HTTPSRedirectMiddleware locally:
```shell
export DEBUG=True
```

Run the application
```shell
./start.sh
```

OpenAPI schema is available at:
```
http://127.0.0.1:8000/latest/docs
http://127.0.0.1:8000/latest/redoc
http://127.0.0.1:8000/latest/openapi.json
```

### List of the environmental variables used in project:


| Variable               | Default value     | Description                                                                         |
|------------------------|-------------------|-------------------------------------------------------------------------------------|
| POSTGRES_DB            | aviauth           |                                                                                     |
| POSTGRES_HOST          | 127.0.0.1         |                                                                                     |
| POSTGRES_PORT          | 5432              |                                                                                     |
| POSTGRES_USER          |                   |                                                                                     |
| POSTGRES_PASSWORD      |                   |                                                                                     |
| DEBUG                  | False             | If set to True, logs to console SQL queries and disable HTTPSRedirectMiddleware.    |
| SECRET_KEY             |                   |                                                                                     |


# Hints & Tips

### Commits

This project is following [Conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) specification
and [Trunk-based development](https://www.atlassian.com/continuous-delivery/continuous-integration/trunk-based-development)
flow.

Common commit types:

`feat:`
`fix:`
`build:`
`chore:`
`ci:`
`docs:`
`style:`
`refactor:`
`perf:`
`test:`

Branch name patterns:

`master`
`feature/{feature_name}`
`bugfix/{hotfix_name}`

#### Before committing, ensure that you have:

* installed pre-commit hook (`pre-commit install`)
* included all the changes to the [CHANGELOG.md](CHANGELOG.md) under
`Unreleased` section
* created a branch according to the pattern described above (e.g.: `feature/{feature_name}`)
* put a correct commit type according to presented above (e.g.: `feat: add new feature`)
* put `!` after the commit type (e.g.: `feat!: and new breaking change`) and place
`BRAKING CHANGE:` in the beginning of commit body (optional)
if you are introducing breaking changes

#### Before the release:

* Set proper release version in the [CHANGELOG.md](CHANGELOG.md)
* Ensure that [CHANGELOG.md](CHANGELOG.md) content is up-to-date
* Set the release date in the [CHANGELOG.md](CHANGELOG.md)
* Update `V{current_major}_VERSION` parameter in [config](app/core/config.py)
according to the releasing one or create a new one if `BREAKING CHANGES`
took place
* Check that `LATEST_VERSION` is pointing to the correct major version parameter in
[config](app/core/config.py)
* After `master` branch is up-to-date, create a release on GitHub including the
latest release notes

###Alembic

Initialize Alembic:
```shell
alembic init -t async migrations
```

Make migration:
```shell
alembic revision --autogenerate -m "migration massage"
```

Apply migration:
```shell
alembic upgrade head
```

Downgrade 1 revision:
```shell
alembic downgrade -1
```

### Security

Generate random secret key:
```shell
openssl rand -hex 32
```


# Overview

This project provides authentication and authorization service for external applications.
It allows using other applications without storing any user data in them.

Keeping all the data in a single place and clear management of access permissions for each
authorized application is a goal of this project.

We can end up with 3 products:
- Independent application for integration with external apps
- Open source authentication microservice ready-to-use
- Pip package for FASTApi


### Problem Statement

Most of the web application require registration or authentication using external services 
(as google or facebook), that have almost unlimited access to users data. Easy management of 
applications and permissions they have is something that missing.
Personal data is not personal anymore.


### Proposed Solution

Implement a service that keeps all the user's data encrypted and providing it
to external applications based on configured permissions.


# Success Criteria

- Running application in production environment
- Allowing to perform any user story described below
- Compliance with all requirements of the project


# User Stories

- Register accounts
- CRUD operations on user data
- Several emails / phone numbers / avatars per user
- Register / Manage applications
- Generate authentication tokens (multiple tokens per app)
- Manage permissions per app
- List / Delete active tokens
- See access history
- [2FA](https://en.wikipedia.org/wiki/Multi-factor_authentication)


# Scope

- Only creation of the API is the scope of the project. No frontend or other UI is required.
- A service for checking username availability in some popular applications
can be implemented in future iterations.


### Requirements

- [FastAPI](https://fastapi.tiangolo.com) web framework
- [SQLAlchemy](https://www.sqlalchemy.org) ORM
- Test coverage with [pytest](https://docs.pytest.org/en/6.2.x/contents.html)
- [OpenAPI](https://swagger.io/specification/) specification
- [GitHub actions](https://github.com/features/actions) as CI/CD tool
- [AWS](http://aws.amazon.com) infrastructure
- Managing infrastructure with [Terraform](https://www.terraform.io)
- [Trunk-based development](https://www.atlassian.com/continuous-delivery/continuous-integration/trunk-based-development)
- [Open Source](https://opensource.org)
- [REST](https://restfulapi.net) API
- [PostgreSQL](https://www.postgresql.org)


### Non-Requirements

- 

### Conventions 

- [Semantic Versioning](https://semver.org/)
- [Changelog](https://keepachangelog.com/en/1.1.0/)
- [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/)


### Under discussion

- Blockchain usage
- Data encryption
- Project structure
- [Clean architecture](https://breadcrumbscollector.tech/python-the-clean-architecture-in-2021/)
- Automatic changelog generation
- Automatic versioning / tagging


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Changelog Convention](https://keepachangelog.com/en/1.0.0/).
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.2.0](https://github.com/julleks/aviauth-api/compare/0.1.0...0.2.0) - 2021-12--27

### Added 

* Specify project requirements by [@rozumalex](https://github.com/rozumalex)
* Configure CI workflow to run tests with [pytest](https://docs.pytest.org/en/6.2.x/) using project versions
  of python and poetry on both Linux and macOS by [@rozumalex](https://github.com/rozumalex)
* Add a step to CI workflow to upload [coverage report](https://pytest-cov.readthedocs.io/en/latest/) to [codecov.io](https://app.codecov.io/) by [@rozumalex](https://github.com/rozumalex)
* Install pre-commit hook: [check-builtin-literals](https://github.com/pre-commit/pre-commit-hooks#check-builtin-literals) by [@rozumalex](https://github.com/rozumalex)
* Install pre-commit hook: [check-docstring-first](https://github.com/pre-commit/pre-commit-hooks#check-docstring-first) by [@rozumalex](https://github.com/rozumalex)
* Install pre-commit hook: [check-yaml](https://github.com/pre-commit/pre-commit-hooks#check-yaml) by [@rozumalex](https://github.com/rozumalex)
* Install pre-commit hook: [check-toml](https://github.com/pre-commit/pre-commit-hooks#check-toml) by [@rozumalex](https://github.com/rozumalex)
* Install pre-commit hook: [detect-private-key](https://github.com/pre-commit/pre-commit-hooks#detect-private-key) by [@rozumalex](https://github.com/rozumalex)
* Install pre-commit hook: [end-of-file-fixer](https://github.com/pre-commit/pre-commit-hooks#end-of-file-fixer) by [@rozumalex](https://github.com/rozumalex)
* Create an entrypoint for [FastAPI application](https://fastapi.tiangolo.com) by [@rozumalex](https://github.com/rozumalex)
* Create dependency function to start database session by [@rozumalex](https://github.com/rozumalex)
* Create configuration class by [@rozumalex](https://github.com/rozumalex)
* Install [SQLModel](https://sqlmodel.tiangolo.com/features/) package by [@rozumalex](https://github.com/rozumalex)
* Setup [Alembic](https://alembic.sqlalchemy.org/en/latest/) for migrations by [@rozumalex](https://github.com/rozumalex)
* Install [structlog](https://www.structlog.org/en/stable/) by [@rozumalex](https://github.com/rozumalex)
* Add a list of allowed [CORS origins](https://fastapi.tiangolo.com/tutorial/cors/) by [@rozumalex](https://github.com/rozumalex)
* Configure application for versioning by [@rozumalex](https://github.com/rozumalex)
* Integrate documentation with [readthedocs](https://docs.readthedocs.io/en/stable/tutorial/) by [@rozumalex](https://github.com/rozumalex)
* Implement API for authentication by [@rozumalex](https://github.com/rozumalex)
* Implement API for user registration by [@rozumalex](https://github.com/rozumalex)
* Implement API for retrieving user profile by [@rozumalex](https://github.com/rozumalex)
* Implement basic API for applications registration by [@rozumalex](https://github.com/rozumalex)
* Implement basic scope-based permissions by [@rozumalex](https://github.com/rozumalex)
* Add HTTPSRedirectMiddleware by [@rozumalex](https://github.com/rozumalex)
* Add TrustedHostMiddleware by [@rozumalex](https://github.com/rozumalex)

## [0.1.0](https://github.com/julleks/aviauth-api/commits/0.1.0) - 2021-12-08

### Added

* Setup [poetry](https://python-poetry.org) environment manager by [@rozumalex](https://github.com/rozumalex)
* Install [pre-commit](https://pre-commit.com) by [@rozumalex](https://github.com/rozumalex)
* Install pre-commit hook: [black](https://github.com/psf/black) by [@rozumalex](https://github.com/rozumalex)
* Install pre-commit hook: [isort](https://github.com/timothycrosley/isort) by [@rozumalex](https://github.com/rozumalex)
* Install pre-commit hook: [flake-8](https://flake8.pycqa.org/en/latest/) by [@rozumalex](https://github.com/rozumalex)
* Install pre-commit hook: [commitizen](https://commitizen-tools.github.io/commitizen/) by [@rozumalex](https://github.com/rozumalex)
* Configure [dependabot](https://help.github.com/github/administering-a-repository/configuration-options-for-dependency-updates) by [@rozumalex](https://github.com/rozumalex)


app/core/config.py
import os
from typing import Any, Dict, List, Optional, Union

from pydantic import AnyHttpUrl, BaseSettings, PostgresDsn, validator


class AsyncPostgresDsn(PostgresDsn):
    allowed_schemes = PostgresDsn.allowed_schemes
    allowed_schemes.add("postgresql+asyncpg")


class Settings(BaseSettings):
    PROJECT_NAME: str = "aviauth-api"

    SECRET_KEY: str = os.getenv("SECRET_KEY", "dev")
    TOKEN_HASH_ALGORITHM: str = "HS256"
    ACCESS_TOKEN_LIFETIME: int = 30 * 60
    REFRESH_TOKEN_LIFETIME: int = 10 * 60 * 60
    API_URL: str = "http://localhost:8000"
    TOKEN_PATH: str = "auth/token"
    TOKEN_URL: str = None

    @validator("TOKEN_URL", pre=True)
    def token_url(cls, v: Optional[str], values: Dict[str, Any]) -> str:
        if v:
            return v
        return f"{values.get('API_URL')}/latest/{values.get('TOKEN_PATH')}"

    V0_VERSION = (0, 2, 0)
    LATEST_VERSION = V0_VERSION

    OAUTH2_SCOPES = {
        "user:read": "Read access to Identity information.",
        "user:update": "Write access to Identity information.",
        "user:full": "Full access to Identity information.",
        "applications:read": "Read access to Applications list.",
        "applications:create": "Create access to Applications list.",
        "applications:update": "Update access to Applications list.",
        "applications:full": "Full access to Applications list.",
    }

    MAX_DB_CONNECTION_RETRIES = 5 * 60  # time in seconds
    DB_CONNECTION_RETRY_WAIT_SECONDS = 5

    DEBUG: bool = os.getenv("DEBUG", False)

    USE_TZ: bool = True

    @validator("BACKEND_CORS_ORIGINS", pre=True)
    def assemble_cors_origins(cls, v: Union[str, List[str]]) -> Union[List[str], str]:
        if isinstance(v, str) and not v.startswith("["):
            return [i.strip() for i in v.split(",")]
        elif isinstance(v, (list, str)):
            return v
        raise ValueError(v)

    # https://fastapi.tiangolo.com/tutorial/cors/
    # https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS
    BACKEND_CORS_ORIGINS: List[AnyHttpUrl] = []

    # A list of domain names that should be allowed as hostnames.
    # Wildcard domains such as *.example.com are supported for matching subdomains
    # to allow any hostname either use allowed_hosts=["*"].
    ALLOWED_HOSTS: List[str] = []

    @validator("DATABASE_URL", pre=True)
    def assemble_db_connection(cls, v: Optional[str], values: Dict[str, Any]) -> Any:
        if isinstance(v, str):
            return v
        return AsyncPostgresDsn.build(
            scheme="postgresql+asyncpg",
            user=values.get("POSTGRES_USER"),
            password=values.get("POSTGRES_PASSWORD"),
            host=values.get("POSTGRES_HOST"),
            port=values.get("POSTGRES_PORT"),
            path=f"/{values.get('POSTGRES_DB')}",
        )

    POSTGRES_HOST: str = os.getenv("POSTGRES_HOST", "127.0.0.1")
    POSTGRES_PORT: str = os.getenv("POSTGRES_PORT", "5432")
    POSTGRES_USER: str = os.getenv("POSTGRES_USER")
    POSTGRES_PASSWORD: str = os.getenv("POSTGRES_PASSWORD")
    POSTGRES_DB: str = os.getenv("POSTGRES_DB", "aviauth")

    DATABASE_URL: Optional[AsyncPostgresDsn] = None


settings = Settings()
